go/updater/trainingJobUpdater.go

```diff
@@ -48,34 +56,423 @@ type TrainingJobUpdater struct {
+}
+
+func (updater *TrainingJobUpdater) releaseMaster() error {
+	return updater.releaseResource(padv1.MASTER)
+}
+
+func (updater *TrainingJobUpdater) releasePserver() error {
+	return updater.releaseResource(padv1.PSERVER)
+}
+
+func (updater *TrainingJobUpdater) releaseTrainer() error {
+	labels := Labels(map[string]string{
+		"paddle-job": updater.job.Name,
+	})
+	selector, _ := labels.LabelsParser()
+	options := v1.ListOptions{
+		LabelSelector: selector,
+	}
+
+	return updater.kubeClient.CoreV1().Pods(updater.job.Namespace).DeleteCollection(&v1.DeleteOptions{}, options)
+}
+
+func (updater *TrainingJobUpdater) deleteTrainingJob() error {
+	fault := false
+
+	log.Infof("Start to delete TrainingJob namespace=%v name=%v", updater.job.Namespace, updater.job.Name)
+
+	if updater.job.Spec.FaultTolerant {
+		log.Infof("Release master, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+		if err := updater.releaseMaster(); err != nil {
+			log.Error(err.Error())
+			fault = true
+		}
+	}
+
+	log.Infof("Release pserver, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+	if err := updater.releasePserver(); err != nil {
+		log.Error(err.Error())
+		fault = true
+	}
+
+	if updater.job.Spec.FaultTolerant {
+		log.Infof("Deleting TrainingJob matadata, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Master.ReplicaSpec.Name)
+		if err := updater.kubeClient.ExtensionsV1beta1().ReplicaSets(updater.job.Namespace).Delete(updater.job.Spec.Master.ReplicaSpec.Name, &v1.DeleteOptions{}); err != nil {
+			log.Error("delete master replicaset error: ", err.Error())
+			fault = true
+		}
+	}
+
+	log.Infof("Deleting TrainingJob matadata, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Pserver.ReplicaSpec.Name)
+	if err := updater.kubeClient.ExtensionsV1beta1().ReplicaSets(updater.job.Namespace).Delete(updater.job.Spec.Pserver.ReplicaSpec.Name, &v1.DeleteOptions{}); err != nil {
+		log.Error("delete pserver replicaset error: ", err.Error())
+		fault = true
+	}
+
+	log.Infof("Deleting TrainingJob matadata, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+	if err := updater.kubeClient.BatchV1().Jobs(updater.job.Namespace).Delete(updater.job.Spec.Trainer.ReplicaSpec.Name, &v1.DeleteOptions{}); err != nil {
+		log.Error("delete trainer replicaset error: ", err.Error())
+		fault = true
+	}
+
+	log.Infof("Release trainer, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+	if err := updater.releaseTrainer(); err != nil {
+		log.Error("release trainer  error: ", err.Error())
+		fault = true
+	}
+
+	log.Infof("End to delete TrainingJob namespace=%v name=%v", updater.job.Namespace, updater.job.Name)
+
+	if fault {
+		return fmt.Errorf("delete resource error namespace=%v name=%v", updater.job.Namespace, updater.job.Name)
```

Comment:
This may lose the original k8s api call error message, return the error object from `releaseXXX` call.